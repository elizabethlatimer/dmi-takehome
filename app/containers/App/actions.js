/*
 *
 * Home actions
 *
 */

import {
  LOAD_QUOTES,
  LOAD_QUOTES_SUCCESS,
  LOAD_QUOTES_ERROR,
  ADD_NEW_QUOTE,
  SEND_QUOTE,
} from './constants';

export function loadQuotes() {
  return {
    type: LOAD_QUOTES,
  };
}

export function quotesLoaded(quotes) {
  return {
    type: LOAD_QUOTES_SUCCESS,
    quotes,
  };
}

export function quoteLoadingError(error) {
  return {
    type: LOAD_QUOTES_ERROR,
    error,
  };
}

export function sendQuote() {
  return {
    type: SEND_QUOTE,
  };
}

export function addNewQuote(quote) {
  return {
    type: ADD_NEW_QUOTE,
    quote,
  };
}
