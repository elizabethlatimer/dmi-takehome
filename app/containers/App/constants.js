/*
 *
 * App constants
 *
 */

export const LOAD_QUOTES = 'app/Home/LOAD_QUOTES';
export const LOAD_QUOTES_SUCCESS = 'app/Home/LOAD_QUOTES_SUCCESS';
export const LOAD_QUOTES_ERROR = 'app/Home/LOAD_QUOTES_ERROR';
export const ADD_NEW_QUOTE = 'app/Home/ADD_NEW_QUOTE';
export const SEND_QUOTE = 'app/Home/SEND_QUOTE';
