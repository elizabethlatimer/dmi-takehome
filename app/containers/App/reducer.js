/*
 *
 * App reducer
 *
 */
import produce from 'immer';
import {
  LOAD_QUOTES,
  LOAD_QUOTES_SUCCESS,
  LOAD_QUOTES_ERROR,
  ADD_NEW_QUOTE,
} from './constants';

export const initialState = {
  loading: true,
  errors: false,
  quotes: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_QUOTES:
        draft.loading = true;
        draft.errors = false;
        break;

      case LOAD_QUOTES_SUCCESS:
        draft.quotes = action.quotes;
        draft.loading = false;
        break;

      case LOAD_QUOTES_ERROR:
        draft.errors = action.error;
        draft.loading = false;
        break;

      case ADD_NEW_QUOTE:
        draft.quotes = [action.quote, ...state.quotes];
        break;
    }
  });

export default appReducer;
