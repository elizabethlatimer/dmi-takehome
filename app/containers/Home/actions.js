/*
 *
 * Home actions
 *
 */

import {
  LOAD_QUOTES,
  LOAD_QUOTES_SUCCESS,
  LOAD_QUOTES_ERROR,
} from './constants';

export function loadQuotes() {
  return {
    type: LOAD_QUOTES,
  };
}

export function quotesLoaded(quotes) {
  return {
    type: LOAD_QUOTES_SUCCESS,
    quotes,
  };
}

export function quoteLoadingError(error) {
  return {
    type: LOAD_QUOTES_ERROR,
    error,
  };
}
