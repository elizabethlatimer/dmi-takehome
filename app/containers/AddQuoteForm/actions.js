/*
 *
 * AddQuoteForm actions
 *
 */

import { TEMP_QUOTE, CLEAR_TEMP_QUOTE } from './constants';

export function setTempQuote(quote) {
  return {
    type: TEMP_QUOTE,
    newQuote: quote,
  };
}

export function clearTempQuote() {
  return {
    type: CLEAR_TEMP_QUOTE,
  };
}
