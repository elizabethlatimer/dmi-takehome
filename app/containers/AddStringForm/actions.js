/*
 *
 * AddStringForm actions
 *
 */

import { POST_QUOTE } from './constants';

export function postQuote() {
  return {
    type: POST_QUOTE,
  };
}
