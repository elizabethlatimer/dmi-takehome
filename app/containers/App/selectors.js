import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    substate => substate.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    substate => substate.error,
  );

const makeSelectQuotes = () =>
  createSelector(
    selectGlobal,
    substate => substate.quotes,
  );

export {
  makeSelectLocation,
  makeSelectLoading,
  makeSelectError,
  makeSelectQuotes,
};
