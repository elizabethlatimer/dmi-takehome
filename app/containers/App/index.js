/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Switch, Route } from 'react-router-dom';

import Home from '../Home/Loadable';
import Nav from '../../components/Nav';
import AddQuoteForm from '../AddQuoteForm';
import NotFoundPage from '../NotFoundPage/Loadable';

import GlobalStyle from '../../global-styles';

export default function App() {
  return (
    <div>
      <Nav />
      <Switch>
        <Route exact path="/" component={Home} />
        <Route exact path="/new" component={AddQuoteForm} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </div>
  );
}
