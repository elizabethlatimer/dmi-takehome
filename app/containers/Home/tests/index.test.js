/**
 *
 * Tests for Home
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { Home, mapDispatchToProps } from '../index';
import { loadQuotes } from '../../App/actions';
import configureStore from '../../../configureStore';

const STRINGS = [
  { quote: 'Be yourself; everyone else is already taken.', id: 0 },
  { quote: 'This above all: to thine own self be true.', id: 1 },
  {
    quote: 'If you cannot do great things, do small things in a great way.',
    id: 2,
  },
  { quote: 'Strive not to be a success, but rather to be of value.', id: 3 },
  {
    quote: 'Do not let what you cannot do interfere with what you can do.',
    id: 4,
  },
];

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <Home loading={false} errors={false} quotes={STRINGS} />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch the quotes on first render if quotes do not exist', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <Home loading quotes={false} onGetQuotes={submitSpy} />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).toHaveBeenCalled();
  });

  it('should not call onGetQuotes if there are quotes in props', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <Home loading quotes={STRINGS} onGetQuotes={submitSpy} />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onGetQuotes', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onGetQuotes).toBeDefined();
      });

      it('should dispatch loadQuotes when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onGetQuotes();
        expect(dispatch).toHaveBeenCalledWith(loadQuotes());
      });
    });
  });
});
