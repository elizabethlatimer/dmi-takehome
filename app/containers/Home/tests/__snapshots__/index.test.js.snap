// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div>
  
  <blockquote
    class="Blockquote-f975vm-0 gRwfUI"
  >
    Be yourself; everyone else is already taken.
  </blockquote>
  <blockquote
    class="Blockquote-f975vm-0 gRwfUI"
  >
    This above all: to thine own self be true.
  </blockquote>
  <blockquote
    class="Blockquote-f975vm-0 gRwfUI"
  >
    If you cannot do great things, do small things in a great way.
  </blockquote>
  <blockquote
    class="Blockquote-f975vm-0 gRwfUI"
  >
    Strive not to be a success, but rather to be of value.
  </blockquote>
  <blockquote
    class="Blockquote-f975vm-0 gRwfUI"
  >
    Do not let what you cannot do interfere with what you can do.
  </blockquote>
</div>
`;
